### PR TITLE
Simplify policy for Boolean term skolems

### PR DESCRIPTION
**Pull request by ajreynol**
This PR cleans up our policy for purification skolems.

In particular, we have been using purification for two purposes:
(1) *Boolean* purification skolems indicate that the variables should be treated as theory atoms for theory combination. This meant that `Theory::theoryOf` checked whether a variable has been marked as a purification skolem, thus making it belong to THEORY_UF.
(2) purification skolems are used as a way of marking the definition of internally introduced variables, making our proof system trivially justify the replacement of terms by their purification skolem.

These two use cases clashed. In particular, if one introduced a purification skolem for any term in a logic that did not include UF, a logic exception would occur.  Moreover this meant that using a purification skolem in a subsolver would be taken out of context as a theory atom.

This makes `Theory::theoryOf` independent of whether a variable is a purification skolem, and instead tracks this set explicitly in `Env`.

This will enable more use cases of purification skolems to justify proofs of preprocessing.

### CODE REVIEW

File: src/smt/env.cpp [hunk 1]
```diff
@@ -282,4 +289,20 @@ theory::quantifiers::OracleChecker* Env::getOracleChecker() const
...
+  d_boolTermSkolems.insert(k);
+}
+
+bool Env::isBooleanTermSkolem(const Node& k) const
+{
+  // optimization: check whether k is a variable
+  if (!k.isVar())
+  {
+    return false;
+  }
```
Comment (yoni206) [inline src/smt/env.cpp:304]: why can't variables be boolean term skolems?

File: src/smt/env.h [hunk 2]
```diff
@@ -279,6 +280,26 @@ class Env
...
   theory::quantifiers::OracleChecker* getOracleChecker() const;
 
+  /**
+   * Register Boolean term skolem. This registers that k is a Boolean variable
+   * that should be treated as a theory atom. This impacts theoryOf, where
+   * Boolean term skolems belong to THEORY_UF, not THEORY_BOOL.
+   *
+   * This method is called by the "remove term formula" pass, which recognizes
+   * when Boolean terms occur in term positions, which are relevant for
+   * theory combination.
```
Comment (yoni206) [inline src/smt/env.h:290]: Not sure if this should be here, as this function can technically be used in other places too.

File: src/smt/env.cpp (hunk 1 above)
Comment (ajreynol) [reply src/smt/env.cpp:304]: There is a `!`, i.e. only variables are Boolean term skolems.

File: src/smt/env.h (hunk 2 above)
Comment (ajreynol) [reply src/smt/env.h:290]: I think it is helpful to state the main use case, but indeed the method can be used in other places.

**Review by yoni206 — APPROVED** [review]
Thanks for the clarifications

**Pull request merged**